src/lir/lir_type.rs

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fmt;
+
+use crate::lir::expressions::Operand;
+use crate::sema::ast;
+use crate::sema::ast::ArrayLength;
+
+/// A struct type definition that is similar to the one in ast.rs,
+/// extended with a Vector type, as we need a lower level representation of
+/// String and DynamicBytes
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum StructType {
+    UserDefined(usize),
+    SolAccountInfo,
+    SolAccountMeta,
+    SolParameters,
+    ExternalFunction,
+    /// Vector is used here to represent String and DynamicBytes
+    Vector(Box<Type>),
+}
+
+/// A struct that contains the AST type and the LIR type.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct LIRType {
+    pub ast_type: ast::Type,
+    pub lir_type: Type,
+}
+
+/// Types for LIR. Some types present in the AST are not present here, as they
+/// are lowered to other types. See the `lower_ast_type` function in the `lir::converter::Converter`.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Type {
+    Bool,
+    Int(u16),
+    Uint(u16),
+    Bytes(u8),
+    Ptr(Box<Type>),
+    StoragePtr(bool, Box<Type>),
+    Function {
+        params: Vec<Type>,
+        returns: Vec<Type>,
+    },
```

Comment:
Surely Function does not need any fields

Reply:
A function pointer in LLVM needs the argument and return types:

https://github.com/hyperledger/solang/blob/92a6c6c89d3002ba6b80c85d8b71a9c5fc82b80b/src/emit/binary.rs#L962-L967

If we have a function pointer, we do need those fields.